public/API/API_GetGameProgression.php

```diff
@@ -0,0 +1,177 @@
+        ->whereNull('unlocker_id')
+        ->get();
+    if ($unlocks->count() === 0) {
+        continue;
+    }
+
+    $achievementSessionStart = $resets[$playerId] ?? $coreSet?->achievements_first_published_at;
+
+    $sessionQuery = PlayerSession::query()
+        ->where('user_id', $playerId)
+        ->where('game_id', $game->ID)
+        ->when($achievementSessionStart, fn ($q) => $q->where('rich_presence_updated_at', '>', $achievementSessionStart))
+        ->select(['created_at', 'duration', 'rich_presence_updated_at']);
+
+    $elapsed = 0;
+    foreach ($sessionQuery->orderBy('rich_presence_updated_at')->get() as $session) {
+        $sessionStart = $achievementSessionStart ? max($achievementSessionStart, $session->created_at) : $session->created_at;
+        $sessionEnd = max($session->rich_presence_updated_at, $session->created_at->addMinutes($session->duration));
+
+        foreach ($unlocks as $unlock) {
+            if ($unlock->unlocked_at && $unlock->unlocked_at->between($sessionStart, $sessionEnd)) {
+                $unlock_times[$unlock->achievement_id][] =
+                    $unlock->unlocked_at->diffInSeconds($sessionStart, true) + $elapsed;
+            }
+            if ($unlock->unlocked_hardcore_at && $unlock->unlocked_hardcore_at->between($sessionStart, $sessionEnd)) {
+                $unlock_hardcore_times[$unlock->achievement_id][] =
+                    $unlock->unlocked_hardcore_at->diffInSeconds($sessionStart, true) + $elapsed;
+            }
+        }
+
+        $elapsed += $sessionEnd->diffInSeconds($sessionStart, true);
+    }
+}
+
+// ===== summarize achievement metrics =====
+$get_median = function (array $a): int {
+    $length = count($a);
+    if ($length === 0) {
+        return 0;
+    }
+
+    $values = array_values($a);
+    sort($values);
+
+    $index = floor($length / 2);
+    if (($length % 2) == 1) {
+        return $values[$index];
+    }
+
+    return (int) round(($values[$index - 1] + $values[$index]) / 2);
+};
+
+foreach ($achievements as $achievement) {
+    $response['Achievements'][] = [
+        'ID' => $achievement->ID,
+        'Title' => $achievement->Title,
+        'Description' => $achievement->Description,
+        'Points' => $achievement->Points,
+        'TrueRatio' => $achievement->TrueRatio,
+        'Type' => $achievement->type,
+        'BadgeName' => $achievement->BadgeName,
+        'NumAwarded' => $achievement->unlocks_total,
+        'NumAwardedHardcore' => $achievement->unlocks_hardcore_total,
+        'TimesUsedInUnlockMedian' => count($unlock_times[$achievement->ID]),
+        'TimesUsedInHardcoreUnlockMedian' => count($unlock_hardcore_times[$achievement->ID]),
+        'MedianTimeToUnlock' => $get_median($unlock_times[$achievement->ID]),
+        'MedianTimeToUnlockHardcore' => $get_median($unlock_hardcore_times[$achievement->ID]),
+    ];
+}
+
+usort($response['Achievements'], fn ($a, $b) => $a['MedianTimeToUnlockHardcore'] - $b['MedianTimeToUnlockHardcore']);
```

Comment:
Potential logic error: should the field used for `$a` and `$b` respect the `$preferHardcore` flag?

Reply:
`$preferHardcore` only drives the behavior of the threshold query (for determining who the last 100 players were). The results still contain data for both hardcore and softcore.